service/dap/daptest/client.go

```diff
@@ -49,30 +49,96 @@ func (c *Client) send(request dap.Message) {
 	dap.WriteProtocolMessage(c.conn, request)
 }
 
-// ReadBaseMessage reads and returns a json-encoded DAP message.
-func (c *Client) ReadBaseMessage() ([]byte, error) {
-	message, err := dap.ReadBaseMessage(c.reader)
+func (c *Client) ExpectDisconnectResponse(t *testing.T) *dap.DisconnectResponse {
+	m, err := dap.ReadProtocolMessage(c.reader)
 	if err != nil {
-		fmt.Println("DAP client error:", err)
-		return nil, err
+		t.Error(err)
 	}
-	fmt.Println("[client <- server]", string(message))
-	return message, nil
+	return m.(*dap.DisconnectResponse)
 }
 
-// ReadErrorResponse reads, decodes and validates the result
-// to be an error response. Returns the response or an error.
-func (c *Client) ReadErrorResponse() (dap.Message, error) {
-	response, err := dap.ReadProtocolMessage(c.reader)
+func (c *Client) ExpectErrorResponse(t *testing.T) *dap.ErrorResponse {
+	m, err := dap.ReadProtocolMessage(c.reader)
 	if err != nil {
-		return nil, err
+		t.Error(err)
 	}
-	switch response.(type) {
-	case *dap.ErrorResponse:
-		return response, nil
-	default:
-		return nil, errors.New(fmt.Sprintf("not an ErrorResponse: %#v", response))
+	return m.(*dap.ErrorResponse)
+}
+
+func (c *Client) ExpectContinueResponse(t *testing.T) *dap.ContinueResponse {
+	m, err := dap.ReadProtocolMessage(c.reader)
```

Comment:
I think you would be better served by defining a helper function:

```
func mustReadProtocolMessage(t *testing.T, c *Client) dap.Message {
   m, err := dap.ReadProtocolMessage(c.reader)
   if err != nil {
       t.Error(err)
   }
   return m
}
```

then you wouldn't even need these, you could just `mustReadProtocolMessage(t, c).(*dap.ConcreteResponseType)`

Reply:
Thanks for the suggestion. Just to make sure I understand precisely, so you mean I should use (1) this helper function to simplify the `Expect*` methods, or (2) replace all calls to `Expect*` with this?

E.g., replace:

```
contResp := client.ExpectContinueResponse(t)
```

With:

```
contResp := mustReadProtocolMessage(t, client).(*dap.ContinueResponse)
```

The tradeoff, IMHO is:

* The original has much cleaner call sites in tests, so the tests themselves are easier to read.
* We have some custom logic in some `Expect*` method (e.g. `ExpectInitializeResponse`) and may add more such logic in the future. Removing the helpers means this logic needs to be duplicated at call sites, or have specific helpers just for relevant methods.

On the other hand, your proposal helps remove a bunch of repetitive methods, which is an advantage.

A middle way is (1), replace:

```
func (c *Client) ExpectContinueResponse(t *testing.T) *dap.ContinueResponse {
	m, err := dap.ReadProtocolMessage(c.reader)
	if err != nil {
		t.Error(err)
	}
	return m.(*dap.ContinueResponse)
}
```

By:

```
func (c *Client) ExpectContinueResponse(t *testing.T) *dap.ContinueResponse {
       return c.mustReadProtocolMessage().(*dap.ContinueResponse)
}
```

This keeps all the `Expect*` methods but makes them one-liners, except those where extra logic is needed. The call sites in tests remain short and readable, but the amount of repetitive code is reduced.

WDYT?

Reply:
I was thinking just:

```
contResp := mustReadProtocolMessage(t, client).(*dap.ContinueResponse)
```

but if you think you'll need to add extra logic then keeping the Expect methods makes sense.

Reply:
I like the idea of keeping the Expect methods, but factoring out all the common parts into a mustRead healper.

Reply:
Except for consistency it should be called ExpectProtocolMessage or the other one should be called MustReadFooRequest.

Reply:
Done